src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs

```diff
@@ -142,6 +142,259 @@ private static void ReadDoubleHelper(ref Utf8JsonReader jsonReader, int expected
+        [InlineData("0.1e ", 4, 0, 0, ' ')]
+        [InlineData("01 ", 1, 0, 0, '1')]
+        [InlineData("1a ", 1, 0, 0, 'a')]
+        [InlineData("-01 ", 2, 0, 0, '1')]
+        [InlineData("10.5e ", 5, 0, 0, ' ')]
+        [InlineData("10.5e- ", 6, 0, 0, ' ')]
+        [InlineData("10.5e+ ", 6, 0, 0, ' ')]
+        [InlineData("10.5e-0.2 ", 7, 0, 0, '.')]
+        public static void InvalidJsonNumberVariousSegmentSizes(string input, int expectedBytePositionInLine, int numberOfSuccessfulReads, int expectedConsumed, char invalidChar)
+        {
+            byte[] utf8 = Encoding.UTF8.GetBytes(input);
+
+            var jsonReader = new Utf8JsonReader(utf8);
+            InvalidReadNumberHelper(ref jsonReader, expectedBytePositionInLine, numberOfSuccessfulReads, expectedConsumed, invalidChar);
+
+            ReadOnlySequence<byte> sequence = JsonTestHelper.GetSequence(utf8, 1);
+            jsonReader = new Utf8JsonReader(sequence);
+            InvalidReadNumberHelper(ref jsonReader, expectedBytePositionInLine, numberOfSuccessfulReads, expectedConsumed, invalidChar);
+
+            for (int splitLocation = 0; splitLocation < utf8.Length; splitLocation++)
+            {
+                sequence = JsonTestHelper.CreateSegments(utf8, splitLocation);
+                jsonReader = new Utf8JsonReader(sequence);
+                InvalidReadNumberHelper(ref jsonReader, expectedBytePositionInLine, numberOfSuccessfulReads, expectedConsumed, invalidChar);
+            }
+
+            for (int firstSplit = 0; firstSplit < utf8.Length; firstSplit++)
+            {
+                for (int secondSplit = firstSplit; secondSplit < utf8.Length; secondSplit++)
+                {
+                    sequence = JsonTestHelper.CreateSegments(utf8, firstSplit, secondSplit);
+                    jsonReader = new Utf8JsonReader(sequence);
+                    InvalidReadNumberHelper(ref jsonReader, expectedBytePositionInLine, numberOfSuccessfulReads, expectedConsumed, invalidChar);
+                }
+            }
+        }
+
+        private static void InvalidReadNumberHelper(ref Utf8JsonReader jsonReader, int expectedBytePositionInLine, int numberOfSuccessfulReads, int expectedConsumed, char invalidChar)
+        {
+            try
+            {
+                for (int i = 0; i < numberOfSuccessfulReads; i++)
+                {
+                    Assert.True(jsonReader.Read());
+
+                    long tokenLength = jsonReader.HasValueSequence ? jsonReader.ValueSequence.Length : jsonReader.ValueSpan.Length;
+                    Assert.Equal(expectedConsumed, tokenLength);
+                    Assert.Equal(0, jsonReader.TokenStartIndex);
+                    Assert.Equal(expectedConsumed, jsonReader.BytesConsumed - jsonReader.TokenStartIndex);
+                }
+                Assert.False(jsonReader.Read());
+                Assert.True(false, "Expected JsonException was not thrown for incomplete/invalid JSON payload reading.");
```

Comment:
Don't you have a dedicated Assert.Throws-like helper for Utf8JsonReader?

Reply:
Had to update it to return the expected exception. The helper didn't return it before.

Reply:
I can't use the helper because I need to pass the Utf8JsonReader by ref and then inspect it's mutated state (i.e. BytesConsumed). I don' see how that's feasible here.

Reply:
I am going to leave things as is for now.